### PR TITLE
[spring] Update HTTP to match Request/Response API

### PR DESCRIPTION
**Pull request by AlexAPB**
This updates `yarpchttp` to match the new `Request` and `Response` APIs that
were introduced in #1566.

Each commit is individually reviewable.

Once this lands, the entire `yarpchttp` package will be now buildable and testable.

### CODE REVIEW

File: v2/yarpchttp/handler.go [hunk 1]
```diff
@@ -52,193 +49,197 @@ type handler struct {
 	logger              *zap.Logger
 }
 
-func (h handler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+func (h handler) ServeHTTP(w http.ResponseWriter, hRequest *http.Request) {
```
Comment (kriskowal) [inline v2/yarpchttp/handler.go]: How about httpReq?

File: v2/yarpchttp/handler.go [hunk 2]
```diff
@@ -52,193 +49,197 @@ type handler struct {
...
 	responseWriter := newResponseWriter(w)
```
Comment (kriskowal) [inline v2/yarpchttp/handler.go]: Such inconsistency about abbreviation…

File: v2/yarpchttp/handler.go [hunk 3]
```diff
@@ -52,193 +49,197 @@ type handler struct {
...
-	service := popHeader(req.Header, ServiceHeader)
-	procedure := popHeader(req.Header, ProcedureHeader)
-	bothResponseError := popHeader(req.Header, AcceptsBothResponseErrorHeader) == AcceptTrue
+	service := popHeader(hRequest.Header, ServiceHeader)
+	procedure := popHeader(hRequest.Header, ProcedureHeader)
+	bothResponseError := popHeader(hRequest.Header, AcceptsBothResponseErrorHeader) == AcceptTrue
```
Comment (kriskowal) [inline v2/yarpchttp/handler.go]: Let’s stick with req/res for the YARPC type throughout. We can distinguish the HTTP as httpReq for a good balance of explicitness and brevity.

File: v2/yarpchttp/handler.go [hunk 4]
```diff
@@ -52,193 +49,197 @@ type handler struct {
...
 		defer span.Finish()
-
-		err = yarpctransport.InvokeUnaryHandler(yarpctransport.UnaryInvokeRequest{
-			Context:        ctx,
-			StartTime:      start,
-			Request:        treq,
-			Handler:        spec.Unary(),
-			ResponseWriter: responseWriter,
-			Logger:         h.logger,
+		yResponse, yResponseBuf, err = yarpctransport.InvokeUnaryHandler(yarpctransport.UnaryInvokeRequest{
```
Comment (kriskowal) [inline v2/yarpchttp/handler.go]: res, resBuf?

File: v2/yarpctransport/handler_invoker.go [hunk 5]
```diff
@@ -51,9 +51,7 @@ type StreamInvokeRequest struct {
 
 // InvokeUnaryHandler calls the handler h, recovering panics and timeout errors,
 // converting them to YARPC errors. All other errors are passed through.
-func InvokeUnaryHandler(
-	i UnaryInvokeRequest,
-) (res *yarpc.Response, buf *yarpc.Buffer, err error) {
+func InvokeUnaryHandler(i UnaryInvokeRequest) (res *yarpc.Response, buf *yarpc.Buffer, err error) {
```
Comment (kriskowal) [inline v2/yarpctransport/handler_invoker.go]: I think there’s an opportunity to reduce some package name noise at the call site by making this a Do method on UnaryInvokeRequest, maybe pivoting context into unaryInvokeRequest.Do(ctx).

Comment (AlexAPB) [reply v2/yarpctransport/handler_invoker.go]: added to notes

File: v2/yarpchttp/handler.go [hunk 6]
```diff
@@ -52,193 +49,197 @@ type handler struct {
...
+		// getting the server error string.
+		responseWriter.WriteSystemHeader(BothResponseErrorHeader, AcceptTrue)
+		responseWriter.WriteSystemHeader(ErrorMessageHeader, status.Message())
+		responseWriter.SetResponse(res, resBuf)
 	} else {
-		responseWriter.ResetBuffer()
-		_, _ = fmt.Fprintln(responseWriter, status.Message())
-		responseWriter.AddSystemHeader("Content-Type", "text/plain; charset=utf8")
+		// Set the error message as the response body (not in a header)
+		responseWriter.WriteSystemHeader("Content-Type", "text/plain; charset=utf8")
```
Comment (amckinney) [inline v2/yarpchttp/handler.go]: For my own understanding: do we not need the `Content-Type` in the previous case, even though we are writing to the response?

File: v2/yarpchttp/handler.go [hunk 7]
```diff
@@ -52,193 +49,197 @@ type handler struct {
...
-				responseWriter.AddSystemHeader("Content-Type", contentType)
-			}
-		}
-	}()
 
-	ctx := req.Context()
-	ctx, cancel, parseTTLErr := parseTTL(ctx, treq, popHeader(req.Header, TTLMSHeader))
+	ctx := httpReq.Context()
+	ctx, cancel, parseTTLErr := parseTTL(ctx, req, popHeader(httpReq.Header, TTLMSHeader))
 	// parseTTLErr != nil is a problem only if the request is unary.
```
Comment (amckinney) [inline v2/yarpchttp/handler.go]: Is this comment out-of-date? It looks like we check the `if parseTTLErr != nil` conditional below.

File: v2/yarpchttp/handler.go [hunk 8]
```diff
@@ -52,193 +49,197 @@ type handler struct {
...
 func (rw *responseWriter) Close(httpStatusCode int) {
+	// We MUST write the HTTP status code prior to writing the response body,
+	// since writing to the response body is an implicit 200.
 	rw.w.WriteHeader(httpStatusCode)
-	if rw.buffer != nil {
-		// TODO: what to do with error?
-		_, _ = rw.buffer.WriteTo(rw.w)
-		bufferpool.Put(rw.buffer)
+	if rw.buf != nil {
+		// TODO: log this error?
```
Comment (amckinney) [inline v2/yarpchttp/handler.go]: From "huh" to "maybe"? :P

File: v2/yarpchttp/outbound.go [hunk 9]
```diff
@@ -243,55 +243,65 @@ func (o *Outbound) withCoreHeaders(req *http.Request, treq *yarpc.Request, ttl t
...
-	return req
+	return httpReq
 }
 
-func getYARPCErrorFromResponse(response *http.Response, bothResponseError bool) error {
+// readCloserToBuffer converts a readCloser to a yarpc.Buffer. This attempts to
+// close the given readCloser. This is useful for both inbound and outbound
+// requests.
+//
+// All returned erros are yarpcerror.CodeInternal.
```
Comment (amckinney) [inline v2/yarpchttp/outbound.go]: nit: erros -> errors

File: v2/yarpchttp/handler.go (hunk 6 above)
Comment (AlexAPB) [reply v2/yarpchttp/handler.go]: We already set the content type at the end of `callHandlerI(..)`, so this step is actually stomping on that. In this specific case, we are making the error message the response body (aka not a structured response) so we need to indicate that it's `"text/plain"` and not `json`/`thrift` etc.

File: v2/yarpchttp/handler.go (hunk 8 above)
Comment (AlexAPB) [reply v2/yarpchttp/handler.go]: :p
next time it'll be probably

**Pull request merged**